Remove Bsymbolic

AlexVlx opened the pull request:
This changes the procedure for retrieving bundled code objects to follow 
the one developed for HC2. The symbols which point into the .kernel
section are internalised so as to not be affected by interposition. At
runtime, alongside the local .kernel section, all other such sections
present in linked in shared objects are obtained via ELF introspection.
This removes the need for using the Bsymbolic hack. Unit tests are
corrected to reflect this. The ELFIO library is used for the
introspection, and it is added into the folder which is to host HC2 in
the near future.

AlexVlx committed:
This changes the procedure for retrieving bundled code objects to follow
the one developed for HC2. The symbols which point into the .kernel
section are internalised so as to not be affected by interposition. At
runtime, alongside the local .kernel section, all other such sections
present in linked in shared objects are obtained via ELF introspection.
This removes the need for using the Bsymbolic hack. Unit tests are
corrected to reflect this. The ELFIO library is used for the
introspection, and it is added into the folder which is to host HC2 in
the near future.

AlexVlx committed:
Merge remote-tracking branch 'origin/clang_tot_upgrade' into feature_remove_bsymbolic_requirement_v2

AlexVlx commented:
This is not sufficient for all cases, hence closing it pending resubmission.

whchung commented:
@AlexVlx could you briefly explain what cases are missing in this PR? I was just planning to review this one and it'd be very helpful to let me gain some insights faster.

AlexVlx committed:
This builds upon the prior change and corrects it to make it invariant
by rapport with ordering of library initialisation. At the same time, a
reduction in overhauled capability is introduced, with new HC2
functionality being used only for obtaining the .kernel sections from
shared objects. The shared object iteration is moved to execute within
HSADevice::CreateKernel, which holds all of the HCC changes. A one-time
cost is incurred for the runtime initialisation of the static
structures underpinning the implementation. The spurious
HSAKernel::executable_ data member is removed, since it was not used
anywhere and it made it difficult to construct values of type HSAKernel.
This has been verified to work with some of the trickier cases we are
dealing with, on top of the SharedLibrary unit-tests.

AlexVlx committed:
Merge remote-tracking branch 'origin/clang_tot_upgrade' into feature_remove_bsymbolic_requirement_v2

AlexVlx commented:
@whchung I've sorted things out, if you'd care to review. Thank you.

whchung commented:
Thanks. Let me do this first thing on Monday.

AlexVlx committed:
Merge remote-tracking branch 'origin/clang_tot_upgrade' into feature_remove_bsymbolic_requirement_v2

# Conflicts:
#	lib/hsa/mcwamp_hsa.cpp

AlexVlx committed:
Adopt interface chance in HSAKernel. Note that nullptr is passed as the
owning HSAExecutable for the shared-object case - if at some point
HSAKernel will start using its _executable member this will create
issues.

scchan commented on hc2/headers/functions/hsa_interfaces.hpp:
```diff
@@ -0,0 +1,299 @@
+#include <utility>
+
+inline
+bool operator==(const hsa_agent_t& x, const hsa_agent_t& y)
```
can we add `static` to all the free functions?

AlexVlx replied:
@scchan we could but in this context I think it may not be very useful. Are you concerned about potential ODR violations if someone else decides to make hsa_agent_t EqualityComparable? The downside is that now you technically may get an operator== code drop in all TUs which include that header (an academic issue, agreed). I do not feel strongly either way, and these are temporary...

scchan replied:
It's mainly about potential ODR violations.  We have other headers in hcc containing inlined functions  without `static` and they are included in multiple .cpp files.  That is causing some portability issues with some compilers.  We are looking into fixing them.

AlexVlx replied:
@scchan done:)

AlexVlx committed:
Mark global scope inline functions as static. Take arguments by value,
and not by reference, since they are essentially a single std::uint64_t
wrapped in a struct.

AlexVlx committed:
Merge remote-tracking branch 'origin/clang_tot_upgrade' into feature_remove_bsymbolic_requirement_v2
